packages/block-library/src/heading/edit.js

```diff
@@ -42,10 +42,13 @@ function HeadingEdit( {
 	} );
 
 	const { canGenerateAnchors } = useSelect( ( select ) => {
-		const settings = select( blockEditorStore ).getSettings();
+		const { getGlobalBlockCount, getSettings } = select( blockEditorStore );
+		const settings = getSettings();
 
 		return {
-			canGenerateAnchors: !! settings.__experimentalGenerateAnchors,
+			canGenerateAnchors:
+				!! settings.__experimentalGenerateAnchors ||
+				getGlobalBlockCount( 'core/table-of-contents' ) > 0,
```

Comment:
👋 I'm late to the PR, but has there been a discussion on ways to avoid this coupling between Heading and ToC?

Reply:
^ this was my first inline comment when I started reviewing, and it doesn't really bother me now that I've processed everything, and in the context of #38780.

Reply:
For the record, it was @mtias who [suggested](https://github.com/WordPress/gutenberg/pull/38780#issuecomment-1052129167) auto-enabling the auto-anchor-generation feature of the Heading block when a ToC is present. I figured it was a good compromise to do this, since apparently having auto-anchor-generation always on was causing issues for some people (though I can't recall what exactly). I could easily remove that from this PR, but I added it because people were complaining that a Table of Contents isn't that useful without the links, and there is currently no visual UI option to turn on the auto-anchor-generation.

I figure this dependency is fine, since both blocks exist in the same package, and the Table of Contents block similarly relies on the Heading block to do its stuff. I also already checked to make sure that `getGlobalBlockCount` won't cause an error if the block it is searching for doesn't exist; it simply returns `0` in those cases.

The prior iteration of the ToC block tried pulling headings from the entire document regardless of block, but this proved too convoluted and expensive, since it meant constructing the HTML of the whole post in memory just to pull heading tags from it. It was decided that only supporting the core Heading and Page Break blocks was a far more reasonable solution.